### PR TITLE
Add support for list folding on vectors and identifiers

### PR DESCRIPTION
**Pull request by tohrnii**
Addressing #153. This PR adds support for list folding on vectors and identifiers referring to constant vectors, variable vectors, trace columns or list comprehension.

### CODE REVIEW

File: ir/src/constraints/list_comprehension.rs
Comment (grjte) [inline ir/src/constraints/list_comprehension.rs:201]: I think maybe this should either return a `SemanticError` or be a `debug_assert`. What was your thinking here?

File: ir/src/error.rs
Comment (grjte) [inline ir/src/error.rs:85]: nit: I would change "...List folding..." to "...list folding..."

File: ir/src/constraints/helpers.rs
Comment (grjte) [inline ir/src/constraints/helpers.rs:18]: Should we put this in `list_folding.rs` instead? It would follow the structure of the `tests` more closely

File: ir/src/tests/list_folding.rs
Comment (grjte) [inline ir/src/tests/list_folding.rs:42]: Do we allow the syntax `a+c`? I think we probably shouldn't - it should only be `a+c[i]` that's supported. Since this is passing, maybe there's a bug to fix (and a corresponding new test to add).

**Review by grjte — CHANGES_REQUESTED** [review]
Thanks @tohrnii, this looks good to me overall. A few nits and 2 points

1. I left a comment inline - maybe we have a bug?
2. maybe it would be good to update the integration test or to split the list folding apart from the current list comprehension integration test and add a new integration test for list folding. Wdyt?

File: ir/src/tests/list_folding.rs
Comment (tohrnii) [reply ir/src/tests/list_folding.rs]: Thanks for catching this. I've included a fix in this PR.

File: ir/src/error.rs
Comment (grjte) [inline ir/src/error.rs:104]: nit: I think there might be a better name for this. It's a type mismatch issue, so "single_trace_column_allowed" seems like a misnomer to me. Maybe "invalid_trace_binding" and the error message could be something like "Expected {} to be a binding to a single trace column."? In general, I want to move to the "binding" terminology in the future anyway

It might not be worth changing this now though. Maybe we should just revisit the errors again after this is merged and the refactors are done

File: ir/src/constraints/list_folding.rs
Comment (grjte) [inline ir/src/constraints/list_folding.rs:6]: nit: I'd prefer to import these from super so that we can see everything that is used from `air_script_core` by looking at the `lib.rs` file at the top of the crate

File: ir/src/error.rs
Comment (grjte) [inline ir/src/error.rs:104]: nit: can you update all of these for the new formatting?

e.g.
```
"Symbol type {symbol_type} is not supported for list folding value type {lf_value_type:?}"
```

Also, maybe it would be better to implement `Display` for the listfolding value type than to use debug printing here - but that could be a separate PR so that we can get this merged

**Review by grjte — APPROVED** [review]
Approved, thank you! I left a couple of minor nits inline, but they could be addressed later. I'll leave it up to you which to do now, if any

**Comment by grjte** [comment]
@tohrnii let's wait until #172 is merged before merging this

File: ir/src/error.rs
Comment (tohrnii) [reply ir/src/error.rs]: Thanks. I've changed this for now but agreed we should revisit all the errors again.

Comment (tohrnii) [reply ir/src/error.rs]: I haven't implemented the Display trait yet. Will open a separate PR.

**Pull request merged**
